fdbserver/ptxn/MutableTeamPeekCursor.actor.h

```diff
@@ -0,0 +1,218 @@
+
+	// The function that translates storage team ID to TLogInterface
+	TLogInterfaceByStorageTeamIDFunc getTLogInterfaceByStorageTeamID;
+
+	// Creates new cursor
+	std::shared_ptr<StorageTeamPeekCursor> createCursor(const StorageTeamID& storageTeamID);
+
+	// Refreshes storageTeamIDs by newStorageTeamIDs, add/remove corresponding storage teams
+	bool updateStorageTeams();
+
+	// A flag indicates if the new storage team information is retrieved from remote, stored to newStorageTeamIDs, and
+	// should replace storageTeamIDs
+	bool shouldUpdateStorageTeamCursor;
+
+protected:
+	virtual const VersionSubsequenceMessage& getImpl() const override;
+
+	virtual bool hasRemainingImpl() const override;
+
+public:
+	MutableTeamPeekCursor(const UID serverID_,
+	                      const StorageTeamID& privateMutationStorageTeamID_,
+	                      const TLogInterfaceByStorageTeamIDFunc& getTLogInterfaceByStorageTeamID_)
+	  : BaseClass(), serverID(serverID_), storageServerToTeamIDKey(::storageServerToTeamIdKey(serverID_)),
+	    storageTeamIDs(privateMutationStorageTeamID_), newStorageTeamIDs(privateMutationStorageTeamID_),
+	    getTLogInterfaceByStorageTeamID(getTLogInterfaceByStorageTeamID_), shouldUpdateStorageTeamCursor(false) {
+
+		const auto& privateMutationStorageTeamID = storageTeamIDs.getPrivateMutationsStorageTeamID();
+
+		BaseClass::addCursor(createCursor(privateMutationStorageTeamID));
+
+		TraceEvent("MutableTeamPeekCursorCreated")
+		    .detail("ServerID", serverID.toString())
+		    .detail("PrivateMutationStorageTeamID", privateMutationStorageTeamID.toString());
+	}
+
+	const StorageServerStorageTeams getStorageTeamIDs() const { return storageTeamIDs; }
+};
+
+template <typename BaseClass>
+std::shared_ptr<StorageTeamPeekCursor> MutableTeamPeekCursor<BaseClass>::createCursor(
+    const StorageTeamID& storageTeamID) {
+
+	const auto tLogInterfaces = getTLogInterfaceByStorageTeamID(storageTeamID);
+	ASSERT(tLogInterfaces.size() > 0);
+	return std::make_shared<StorageTeamPeekCursor>(
+	    // NOTE: Do NOT use BaseCalss::getVersion() since it will trigger getImpl
+	    // The new cursor will start peeking *AFTER* current version is completed. The peek will be called immediately
+	    // since hasRemaining will return false after the cursor is added; however the version is one after the current
+	    // so the mutations in the current version will not be included.
+	    BaseClass::currentVersion + 1,
+	    storageTeamID,
+	    tLogInterfaces,
+	    /* pArena */ nullptr,
+	    /* reportEmptyVersion */ true);
+}
+
+template <typename BaseClass>
+bool MutableTeamPeekCursor<BaseClass>::updateStorageTeams() {
+	const auto& newStorageTeams = newStorageTeamIDs.getStorageTeams();
+	const auto& oldStorageTeams = storageTeamIDs.getStorageTeams();
+
+	// NOTE: Do NOT use BaseCalss::getVersion() since it will trigger getImpl
+	TraceEvent("MutableTeamPeekCursorUpdateTeam")
+	    .detail("StorageServerID", serverID.toString())
```

Comment:
`.toString()` can be omitted, i.e.,
```suggestion
	    .detail("StorageServerID", serverID)
```